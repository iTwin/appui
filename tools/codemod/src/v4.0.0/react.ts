/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import type { API, FileInfo, Options } from "jscodeshift";
import { useExtensions } from "../utils/Extensions";

export default function transformer(
  file: FileInfo,
  api: API,
  options: Options
) {
  const j = api.jscodeshift;
  useExtensions(j);

  const root = j(file.source);

  root.rename(
    "@itwin/appui-react:CommonWidgetProps",
    "@itwin/appui-react:Widget"
  );
  root.rename("@itwin/appui-react:WidgetProps", "@itwin/appui-react:Widget");

  return root.toSource(options.printOptions);
}
