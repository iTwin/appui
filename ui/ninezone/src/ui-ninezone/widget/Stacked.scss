/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/style/themecolors";
@import "@bentley/ui-core/lib/ui-core/_z-index";
@import "_variables";
@import "../widgetopacity";

.nz-widget-stacked {
  $min-height: 200px;
  $grip-offset: $nz-resize-grip-size / 2;

  display: flex;
  max-height: 100%;
  position: relative;

  @include nz-widget-opacity;

  > .nz-primary-grip {
    @include uicore-z-index(widget-grip);
  }

  > .nz-height-expander {
    display: flex;
    flex-direction: column;

    &:after {
      content: '';
      flex: 0 1 $min-height;
    }
  }

  > .nz-content-container {
    flex-grow: 1;
    min-width: 0;
    position: relative;

    pointer-events: auto;

    > .nz-content {
      overflow: hidden;
      height: 100%;
      width: 100%;
      background-color: $nz-widget-background-color;
      border: {
        style: solid;
        width: $nz-widget-border-width;
        color: $nz-widget-border-color;
        radius: $nz-widget-border-radius;
      }
      box-sizing: border-box;
    }

    > .nz-secondary-grip {
      @include uicore-z-index(widget-grip);
    }

    > .nz-content-grip {
      @include uicore-z-index(widget-grip);
    }
  }

  > .nz-tabs-column {
    display: flex;
    flex-direction: column;
    position: relative;

    > .nz-tabs {
      display: flex;
    }

    > .nz-tabs-grip-container {
      position: relative;
      flex-basis: $nz-tab-height * 1.5;
      flex-grow: 1;

      > .nz-tabs-grip {
        position: absolute;
        height: 100%;

        @include uicore-z-index(widget-grip);
      }
    }
  }

  > .nz-tab-bar {
    pointer-events: auto;
    position: absolute;
    top: 0;
    left: 0;
    background-color: $buic-background-panel;
    height: 100%;
    width: 100%;
  }

  &.nz-collapsed {
    > .nz-content-container {
      display: none;
    }

    &.nz-bottom-panel-anchor, &.nz-top-panel-anchor {
      display: block;

      > .nz-tabs-column {
        > .nz-tabs-grip-container {
          display: none;
        }
      }

      > .nz-height-expander {
        display: none;
      }

      &.nz-right-anchor {
        > .nz-tabs-column {
          align-items: flex-start;
        }
      }
    }
  }

  &.nz-left-anchor {
    > .nz-tabs-column {
      > .nz-tabs {
        align-items: flex-start;
      }
    }
  }

  &.nz-right-anchor {
    flex-direction: row-reverse;

    > .nz-tabs-column {
      > .nz-tabs {
        align-items: flex-end;
      }
    }
  }

  &.nz-dragged {
    @include uicore-z-index(dragged-widget);
  }

  &.nz-closed {
    height: 100%;
    align-items: center;

    &.nz-bottom-anchor {
      align-items: flex-end;
    }

    > .nz-primary-grip {
      display: none;
    }

    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        flex-basis: 0;

        > .nz-tabs-grip {
          display: none;
        }
      }
    }

    > .nz-content-container {
      display: none;
    }
  }

  &.nz-fill-zone {
    height: 100%;
  }

  &:not(.nz-floating) {
    > .nz-content-container {
      > .nz-content-grip {
        display: none;
      }
    }
  }

  &.nz-bottom-panel-anchor {
    flex-direction: column-reverse;

    > .nz-content-container {
      > .nz-content {
        border-top-left-radius: 0;
      }

      > .nz-content-grip {
        bottom: -$grip-offset;
      }
    }

    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        > .nz-tabs-grip {
          bottom: -$grip-offset;
        }
      }
    }
  }

  &.nz-top-panel-anchor {
    flex-direction: column;

    > .nz-content-container {
      > .nz-content {
        border-bottom-left-radius: 0;
      }

      > .nz-content-grip {
        top: -$grip-offset;
      }
    }

    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        > .nz-tabs-grip {
          top: -$grip-offset;
        }
      }
    }
  }

  &.nz-bottom-panel-anchor, &.nz-top-panel-anchor {
    height: 100%;

    &:not(.nz-collapsed) {
      min-height: $min-height;
    }

    &:not(.nz-fill-zone) {
      max-height: fit-content;
    }

    > .nz-tabs-column {
      flex-direction: row;

      > .nz-tabs {
        flex-direction: row;
      }
    }

    > .nz-height-expander {
      display: none;
    }

    > .nz-content-container {
      min-height: 0;

      > .nz-secondary-grip {
        top: 0;
        right: -$grip-offset;
      }
    }

    > .nz-primary-grip {
      left: -$grip-offset;
    }
  }

  &:not(.nz-bottom-panel-anchor):not(.nz-top-panel-anchor) {
    &.nz-left-anchor {
      > .nz-content-container {
        > .nz-content {
          border-top-right-radius: 0;
        }

        > .nz-content-grip {
          left: -$grip-offset;
        }
      }

      > .nz-tabs-column {
        > .nz-tabs-grip-container {
          > .nz-tabs-grip {
            left: -$grip-offset;
          }
        }
      }
    }

    &.nz-right-anchor {
      > .nz-content-container {
        > .nz-content {
          border-top-left-radius: 0;
        }

        > .nz-content-grip {
          right: -$grip-offset;
        }
      }

      > .nz-tabs-column {
        > .nz-tabs-grip-container {
          > .nz-tabs-grip {
            right: -$grip-offset;
          }
        }
      }
    }

    > .nz-content-container {
      > .nz-secondary-grip {
        bottom: -$grip-offset;
      }

      > .nz-content-grip {
        top: 0;
      }
    }

    > .nz-tabs-column {
      > .nz-tabs {
        flex-direction: column;
      }
    }

    > .nz-primary-grip {
      top: -$grip-offset;
    }
  }
}
