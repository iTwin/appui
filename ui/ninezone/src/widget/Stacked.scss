/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/colors";
@import "../_z-index";
@import "_variables";

.nz-widget-stacked {
  $min-height: 200px;

  display: flex;
  max-height: 100%;
  position: relative;

  $border-width: $nz-widget-border-width;

  > .nz-height-expander {
    display: flex;
    flex-direction: column;

    &:after {
      content: '';
      flex: 0 1 $min-height;
    }
  }

  > .nz-top-grip {
    width: 100%;
    position: absolute;
    top: -($resize-grip-size - $border-width) / 2;
  }

  > .nz-content-area {
    flex-grow: 1;
    min-width: 0;
    border-style: solid;
    border-width: $border-width;
    border-color: $nz-widget-border-color;
    background-color: $nz-widget-background-color;
    position: relative;
    border-bottom-left-radius: $nz-widget-border-radius;
    border-bottom-right-radius: $nz-widget-border-radius;
    pointer-events: auto;

    > .nz-bottom-grip {
      width: 100%;
      position: absolute;
      bottom: -($resize-grip-size - $border-width);
    }

    > .nz-content-grip {
      display: none;
      height: 100%;
      position: absolute;
      top: 0;
    }

    > .nz-settings {
      position: absolute;
      right: -$border-width;
      top: -$border-width;
    }
  }

  > .nz-tabs-column {
    display: flex;
    flex-direction: column;

    > .nz-tabs-grip-container {
      position: relative;
      flex-basis: $nz-tab-height * 1.5;
      flex-grow: 1;

      > .nz-tabs-grip {
        position: absolute;
        height: 100%;
      }
    }
  }

  &.nz-left-anchor {
    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        > .nz-tabs-grip {
          left: -($resize-grip-size + $border-width) / 2;
        }
      }
    }

    > .nz-content-area {
      border-top-left-radius: $nz-widget-border-radius;

      > .nz-settings {
        left: -$border-width;
        border-top-left-radius: $nz-widget-border-radius;
      }

      > .nz-content-grip {
        left: -($resize-grip-size - $border-width);
      }
    }
  }

  &.nz-right-anchor {
    flex-direction: row-reverse;

    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        > .nz-tabs-grip {
          right: -($resize-grip-size + $border-width) / 2;
        }
      }
    }

    > .nz-content-area {
      border-top-right-radius: $nz-widget-border-radius;

      > .nz-settings {
        border-top-right-radius: $nz-widget-border-radius;
      }

      > .nz-content-grip {
        right: -($resize-grip-size - $border-width);
      }
    }
  }

  &.nz-is-dragged {
    @include nz-z-index(dragged-widget);
  }

  &.nz-is-closed {
    height: 100%;
    align-items: center;

    &.nz-bottom-anchor, &.nz-bottom-anchor {
      align-items: flex-end;
    }

    > .nz-top-grip {
      display: none;
    }

    > .nz-tabs-column {
      > .nz-tabs-grip-container {
        flex-basis: 0;

        > .nz-tabs-grip {
          display: none;
        }
      }

      > .nz-tabs {
        > * {
          &.nz-is-active {
            margin-right: unset;
          }
        }
      }
    }

    > .nz-content-area {
      display: none;
    }
  }

  &.nz-fill-zone {
    height: 100%;
  }

  &.nz-is-floating {
    > .nz-content-area {
      > .nz-content-grip {
        display: block;
      }
    }
  }
}
