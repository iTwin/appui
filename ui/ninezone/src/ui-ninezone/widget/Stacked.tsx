/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Widget */

import * as classnames from "classnames";
import * as React from "react";
import { Edge } from "../utilities/Rectangle";
import { WidgetContent } from "./rectangular/Content";
import { ResizeGrip, ResizeDirection } from "./rectangular/ResizeGrip";
import { ResizeHandle } from "./rectangular/ResizeHandle";
import { CommonProps, NoChildrenProps } from "../utilities/Props";
import "./Stacked.scss";

/** Available [[Stacked]] widget horizontal anchors. */
export enum HorizontalAnchor {
  Left,
  Right,
}

/** Available [[Stacked]] widget vertical anchors. */
export enum VerticalAnchor {
  Middle,
  Bottom,
}

/** Helpers for [[HorizontalAnchor]]. */
export class HorizontalAnchorHelpers {
  /** Class name of [[HorizontalAnchor.Left]] */
  public static readonly LEFT_CLASS_NAME = "nz-left-anchor";
  /** Class name of [[HorizontalAnchor.Right]] */
  public static readonly RIGHT_CLASS_NAME = "nz-right-anchor";

  /** @returns Class name of specified [[HorizontalAnchor]] */
  public static getCssClassName(anchor: HorizontalAnchor): string {
    switch (anchor) {
      case HorizontalAnchor.Left:
        return HorizontalAnchorHelpers.LEFT_CLASS_NAME;
      case HorizontalAnchor.Right:
        return HorizontalAnchorHelpers.RIGHT_CLASS_NAME;
    }
  }
}

/** Helpers for [[VerticalAnchor]]. */
export class VerticalAnchorHelpers {
  /** Class name of [[VerticalAnchor.Middle]] */
  public static readonly MIDDLE_CLASS_NAME = "nz-middle-anchor";
  /** Class name of [[VerticalAnchor.Bottom]] */
  public static readonly BOTTOM_CLASS_NAME = "nz-bottom-anchor";

  /** @returns Class name of specified [[VerticalAnchor]] */
  public static getCssClassName(anchor: VerticalAnchor): string {
    switch (anchor) {
      case VerticalAnchor.Middle:
        return VerticalAnchorHelpers.MIDDLE_CLASS_NAME;
      case VerticalAnchor.Bottom:
        return VerticalAnchorHelpers.BOTTOM_CLASS_NAME;
    }
  }
}

/** Properties of [[Stacked]] component. */
export interface StackedProps extends CommonProps, NoChildrenProps {
  /** Content of this widget. */
  content?: React.ReactNode;
  /** Describes if the widget should fill the zone. */
  fillZone?: boolean;
  /** Describes to which side the widget is horizontally anchored. */
  horizontalAnchor: HorizontalAnchor;
  /** Describes if the widget is being dragged. */
  isDragged?: boolean;
  /** Describes if the widget is floating. */
  isFloating?: boolean;
  /** True if widget is open, false otherwise. */
  isOpen?: boolean;
  /** Function called when resize action is performed. */
  onResize?: (x: number, y: number, handle: ResizeHandle, filledHeightDiff: number) => void;
  /** Widget tabs. See: [[Draggable]], [[TabSeparator]], [[Tab]], [[Group]] */
  tabs?: React.ReactNode;
  /** Describes to which side the widget is vertically anchored. Defaults to [[VerticalAnchor.Middle]] */
  verticalAnchor?: VerticalAnchor;
}

/**
 * Stacked widget is used to display multiple tabs and some content.
 * @note Should be placed in [[Zone]] component.
 */
export class Stacked extends React.PureComponent<StackedProps> {
  private _widget = React.createRef<HTMLDivElement>();

  private _getFilledHeightDiff(): number {
    if (this.props.fillZone)
      return 0;

    const widget = this._widget.current;
    if (!widget)
      return 0;

    const heightStyle = widget.style.height;
    const height = widget.clientHeight;

    widget.style.height = "100%";
    const filledHeight = widget.clientHeight;

    widget.style.height = heightStyle;

    const offset = filledHeight - height;
    return offset;
  }

  private _handleTabsGripResize = (x: number) => {
    const filledHeightDiff = this._getFilledHeightDiff();
    switch (this.props.horizontalAnchor) {
      case HorizontalAnchor.Left: {
        this.props.onResize && this.props.onResize(x, 0, Edge.Right, filledHeightDiff);
        break;
      }
      case HorizontalAnchor.Right: {
        this.props.onResize && this.props.onResize(x, 0, Edge.Left, filledHeightDiff);
        break;
      }
    }
  }

  private _handleContentGripResize = (x: number) => {
    const filledHeightDiff = this._getFilledHeightDiff();
    switch (this.props.horizontalAnchor) {
      case HorizontalAnchor.Left: {
        this.props.onResize && this.props.onResize(x, 0, Edge.Left, filledHeightDiff);
        break;
      }
      case HorizontalAnchor.Right: {
        this.props.onResize && this.props.onResize(x, 0, Edge.Right, filledHeightDiff);
        break;
      }
    }
  }

  private _handleTopGripResize = (_x: number, y: number) => {
    const filledHeightDiff = this._getFilledHeightDiff();
    this.props.onResize && this.props.onResize(0, y, Edge.Top, filledHeightDiff);
  }

  private _handleBottomGripResize = (_x: number, y: number) => {
    const filledHeightDiff = this._getFilledHeightDiff();
    this.props.onResize && this.props.onResize(0, y, Edge.Bottom, filledHeightDiff);
  }

  public render() {
    const className = classnames(
      "nz-widget-stacked",
      HorizontalAnchorHelpers.getCssClassName(this.props.horizontalAnchor),
      VerticalAnchorHelpers.getCssClassName(this.props.verticalAnchor === undefined ? VerticalAnchor.Middle : this.props.verticalAnchor),
      !this.props.isOpen && "nz-is-closed",
      this.props.isDragged && "nz-is-dragged",
      this.props.isFloating && "nz-is-floating",
      this.props.fillZone && "nz-fill-zone",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
        ref={this._widget}
      >
        <div className="nz-content-area">
          <WidgetContent
            anchor={this.props.horizontalAnchor}
            content={this.props.content}
          />
          <ResizeGrip
            className="nz-bottom-grip"
            direction={ResizeDirection.NorthSouth}
            onResize={this._handleBottomGripResize}
          />
          <ResizeGrip
            className="nz-content-grip"
            direction={ResizeDirection.EastWest}
            onResize={this._handleContentGripResize}
          />
        </div>
        <div className="nz-tabs-column">
          <div className="nz-tabs">
            {this.props.tabs}
          </div>
          <div className="nz-tabs-grip-container">
            <ResizeGrip
              className="nz-tabs-grip"
              direction={ResizeDirection.EastWest}
              onResize={this._handleTabsGripResize}
            />
          </div>
        </div>
        <div className="nz-height-expander" />
        <ResizeGrip
          className="nz-top-grip"
          direction={ResizeDirection.NorthSouth}
          onResize={this._handleTopGripResize}
        />
      </div>
    );
  }
}
