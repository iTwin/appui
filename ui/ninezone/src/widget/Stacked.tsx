/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Widget */

import * as classnames from "classnames";
import * as React from "react";
import { Edge } from "../utilities/Rectangle";
import Content from "./rectangular/Content";
import ResizeGrip, { ResizeDirection } from "./rectangular/ResizeGrip";
import ResizeHandle from "./rectangular/ResizeHandle";
import CommonProps, { NoChildrenProps } from "../utilities/Props";
import "./Stacked.scss";

/** Available [[Stacked]] widget anchor directions. */
export enum Anchor {
  Left,
  Right,
}

/** Properties of [[Stacked]] component. */
export interface StackedProps extends CommonProps, NoChildrenProps {
  /** Describes to which side the widget is anchored. */
  anchor?: Anchor;
  /** Content of this widget. */
  content?: React.ReactNode;
  /** True if widget is open, false otherwise. */
  isOpen?: boolean;
  /** Function called when resize action is performed. */
  onResize?: (x: number, y: number, handle: ResizeHandle) => void;
  /** One or more tabs/separators. See: [[Draggable]], [[TabSeparator]], [[Tab]] */
  tabs?: React.ReactNode;
}

/**
 * Stacked widget is used to display multiple tabs and some content.
 * @note Should be placed in [[Zone]] component.
 */
// tslint:disable-next-line:variable-name
export const Stacked: React.StatelessComponent<StackedProps> = (props: StackedProps) => {
  const className = classnames(
    "nz-widget-stacked",
    props.anchor === Anchor.Left && "nz-left-anchor",
    !props.isOpen && "nz-is-closed",
    props.className);

  return (
    <div className={className} style={props.style}>
      <div className="nz-content-area">
        <Content
          className="nz-content"
          anchor={props.anchor}
          content={props.content}
        />
        <ResizeGrip
          className="nz-bottom-grip"
          direction={ResizeDirection.NorthSouth}
          onResize={(_x, y) => { props.onResize && props.onResize(0, y, Edge.Bottom); }}
        />
        <ResizeGrip
          className="nz-right-grip"
          direction={ResizeDirection.EastWest}
          onResize={(x) => { props.onResize && props.onResize(x, 0, Edge.Right); }}
        />
      </div>
      <div className="nz-tabs-column">
        <div className="nz-tabs">
          {props.tabs}
        </div>
        <div className="nz-left-grip-container">
          <ResizeGrip
            className="nz-left-grip"
            direction={ResizeDirection.EastWest}
            onResize={(x) => { props.onResize && props.onResize(x, 0, Edge.Left); }}
          />
        </div>
      </div>
      <ResizeGrip
        className="nz-top-grip"
        direction={ResizeDirection.NorthSouth}
        onResize={(_x, y) => { props.onResize && props.onResize(0, y, Edge.Top); }}
      />
    </div>
  );
};

export default Stacked;
