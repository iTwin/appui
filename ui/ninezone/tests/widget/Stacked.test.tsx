/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import { mount, shallow } from "enzyme";
import * as React from "react";

import Stacked from "@src/widget/Stacked";

describe("<Stacked />", () => {
  it("should render", () => {
    mount(<Stacked />);
  });

  it("renders correctly", () => {
    shallow(<Stacked />).should.matchSnapshot();
  });
});
