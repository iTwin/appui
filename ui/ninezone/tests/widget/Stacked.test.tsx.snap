// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stacked /> renders correctly 1`] = `
<div
  className="nz-widget-stacked nz-right-anchor nz-middle-anchor nz-is-closed"
>
  <div
    className="nz-content-area"
  >
    <WidgetContent
      anchor={1}
    />
    <ResizeGrip
      className="nz-bottom-grip"
      direction={1}
      onResize={[Function]}
    />
    <ResizeGrip
      className="nz-content-grip"
      direction={0}
      onResize={[Function]}
    />
  </div>
  <div
    className="nz-tabs-column"
  >
    <div
      className="nz-tabs"
    />
    <div
      className="nz-tabs-grip-container"
    >
      <ResizeGrip
        className="nz-tabs-grip"
        direction={0}
        onResize={[Function]}
      />
    </div>
  </div>
  <div
    className="nz-height-expander"
  />
  <ResizeGrip
    className="nz-top-grip"
    direction={1}
    onResize={[Function]}
  />
</div>
`;
